artiq/firmware/satman/dma.rs

```diff
@@ -0,0 +1,98 @@
+use core::str;
+use board_misoc::csr;
+use alloc::{vec::Vec, collections::btree_map::BTreeMap};
+
+const ALIGNMENT: usize = 64;
+
+#[derive(Debug)]
+struct Entry {
+    trace: Vec<u8>,
+    padding_len: usize,
+    complete: bool
+}
+
+#[derive(Debug)]
+pub struct Manager {
+    entries: BTreeMap<u32, Entry>,
+}
+
+impl Manager {
+    pub fn new() -> Manager {
+        Manager {
+            entries: BTreeMap::new(),
+        }
+    }
+
+    pub fn add(&mut self, id: u32, last: bool, trace: &[u8], trace_len: usize) -> Result<(), &'static str> {
+        let entry = match self.entries.get_mut(&id) {
```

Comment:
There's no ``recording_trace``, only one ``BTree``. If for some reason there's multiple traces written, with their messages interwoven, it would also work.

The ``id`` should be generated by the master, and replaces the name string that was used for identifying traces.